standards/core/rln-contract.md

```diff
@@ -0,0 +1,199 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This specification describes the smart contract that governs RLN memberships, in particular:
+
+- the overall smart contract functionality;
+- contract parameters;
+- the values of those parameters for the initial deployment on a mainnet chain;
+- the way in which the parameters can be modified.
+
+## Background
+
+[Rate-Limiting Nullifiers](https://rate-limiting-nullifier.github.io/rln-docs/) (RLN) is a ZK-based rate-limiting technique used in Waku.
+Before sending messages, users must register in a smart contract.
+They then attach a ZK proof of valid membership to every message.
+Relaying nodes check message validity and drop invalid messages.
+
+A message is considered valid in terms of RLN if:
+- its sender has an active RLN membership; and
+- the sender hasn't exceeded their rate limit within the current epoch.
+
+RLN is managed by a smart contract that keeps track of active memberships.
+Users interact with the contract to register a new membership and
+to obtain Merkle proofs and the tree root,
+which is necessary for proof generation and verification.
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require users to put down a deposit.
+The aim of the deposit initially is purely to protect the system from denial-of-service attacks using bandwidth capping.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+This section is not intended to describe the full contract functionality.
+We only focus on functions required for managing memberships, expiration, deposits, etc.
+The document might later evolve into a full-fledged contract specification.
+
+The RLN contract provides the following functionalities:
+- register a membership in a given tier;
+- extend a membership;
+- withdraw the deposit for an expired membership.
```

Comment:
I'm not sure I agree with

> No one knows when a certain membership will expire

As of now, a membership has an active lifetime of 90 days, followed by a 30-day grace period, during which it can be either extended or "unregistered" (i.e., deposit withdrawn). The smart contract knows whether 90 days have passed since a given membership was registered. I hope this all becomes clear after I add a state transition diagram (WIP).

> we should notify that his/her account has expired

Although the contract itself can't notify anyone, I agree that this would be a useful service from a service provider (I'm thinking about RLNaaS model here).

> but he/she should be able to withdraw his/her funds

This is indeed the case. Note: I didn't mean "for any expired membership", rather "for (that user's) expired memberships".

I'm also not sure what the difference is between "unregistering" and "withdrawing the deposit". In your view, what should happen during "unregistering"?

Reply:
> I'm not sure I agree with
> 
> > No one knows when a certain membership will expire
> 

Okay. By that, I mean that the users usually are not aware of any expiry date. For instance, I don't have a clue when a credit card expires until I get an email from any provider informing me about that.
With that comment, I wanted to bring up the need to inform the user's app about the expiry so that the user's app can inform about it when the expiration either happened or is about to happen.

> I'm also not sure what the difference is between "unregistering" and "withdrawing the deposit". In your view, what should happen during "unregistering"?

I might have misunderstood something. IMHO, the RLN account owner should be the only one allowed to withdraw/unregister the account, in the scenario of account expiration.

Reply:
> the need to inform the user's app about the expiry

I agree with the premise, but do you think this is something a contract should do? The contract itself can't send notifications AFAIU, so some service must track the blockchain events and relay notifications to the user. I agree that this will be useful, it just seems out of scope for a contract spec.

> the RLN account owner should be the only one allowed to withdraw/unregister the account

I agree. However, additionally, the account can "unregister" automatically after its expiration date, if not extended. In that case, its slot in the tree can be overtaken, but the original user's deposit is safe anyway (IMO there should be an extra logic to track unclaimed deposits of overriden expired memberships, letting their users finally withdraw them.) Does this design make sense?

Reply:
> the need to inform the user's app about the expiry

This is not something the contract can do for that matter.

This is a client implementation detail to decide how to communicate to the user what is the RLN membership expire. Depending on the client (mobile app vs cloud node) then it may be done in a different manner.

Reply:
I added a recommendation in the Implementation Suggestion section to save the expiry date locally and remind the user when their membership is about to expire.